knimin/lib/data_access.py

```diff
@@ -2415,6 +2415,144 @@ def set_deposited_ebi(self):
                  WHERE barcode IN %s"""
         self._con.execute(sql, [barcodes])
 
+    def create_study(self, study_id, title=None, alias=None, notes=None):
+        """Creates a study
+
+        Parameters
+        ----------
+        study_id : int
+            Positive if it is a Qiita study ID, negative if others
+        title : str
+        alias : str
+        notes : str
+
+        Returns
+        -------
+        bool
+            True if successful
+
+        Raises
+        ------
+        ValueError
+            If there is error creating the study
+        """
+        with self._con.cursor() as cur:
+            cur.execute('BEGIN')
+            sql = """SELECT study_id, title FROM pm.study WHERE study_id = %s
+                     OR title = %s"""
+            sql_args = [study_id, title or None]
+            qry = self._con.execute_fetchone(sql, sql_args)
+            if qry is not None:
+                raise ValueError('Study ID or title conflicts with exisiting '
+                                 'study %s: %s.' % (qry[0], qry[1]))
+            sql = """INSERT INTO pm.study (study_id, title, alias, notes)
+                     VALUES (%s, %s, %s, %s) RETURNING study_id"""
+            sql_args = [study_id, title or None, alias or None, notes or None]
+            if self._con.execute_fetchone(sql, sql_args) is None:
+                raise ValueError('Creation of study %s failed.' % study_id)
+            cur.execute('COMMIT')
+        return True
+
+    def edit_study(self, study_id, title=None, alias=None, notes=None):
+        """Edits properties of an existing study
+
+        Parameters
+        ----------
+        study_id : int
+        title : str
+        alias : str
+        notes : str
+
+        Returns
+        -------
+        bool
+            True if successful
+
+        Raises
+        ------
+        ValueError
+            If there is error editing the study's properties
+        """
+        with self._con.cursor() as cur:
+            cur.execute('BEGIN')
+            sql = """SELECT * FROM pm.study WHERE study_id = %s LIMIT 1"""
```

Comment:
That's because, for example:
**create_study** wants to check either ID or name exists, and to tell user the conflicting record's name and ID.
**edit_study** wants to check if name exists but its ID is different, and to tell user you are reusing a duplicate name.
**read_study** and **delete_study** only check if name exists.
In the future these commands will also check other relevant tables, e.g., if some samples have been assigned to a study the user wants to delete.
Given these complicated situations, I didn't come up with a generic solution. Actually I would always prefer generic ones if there is any, even the performance is not optimal. Here seems to be a different case.